### PR TITLE
Start permissions documentation

### PR DESCRIPTION
**Pull request by taybenlor**
Adds basic examples from permissions tests, shows how errors will be returned and explains how this could be combined with `get_context` to obtain details from the `Request` object.

## Description

This change adds documentation for the current state of the `BasePermission` class and its behaviour in the system. I am not particularly familiar with Strawberry, but I needed this documentation myself, so I thought a good learning process would be to write it.

To help readers understand the documentation I've explained how the `has_permission` method will return an error, and included details about how you might implement the `has_permission` method. Including suggesting how to add details to the `Info.context` object.

I haven't checked that my approach works, and I don't know if this is the suggested way to do permission. However I figured that just writing something would be a good start and would allow for iteration.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Documentation

## Issues Fixed or Closed by This PR

Not sure, seems like documentation is in meta-issues

## Checklist

Hi, I haven't done any of this! I've just submitted this through the github editor. Would love pointers if there is anything I'm missing. I couldn't see anything in CONTRIBUTING about documentation.

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: docs/guides/permissions.md [hunk 1]
```diff
@@ -4,4 +4,73 @@ title: Permissions
 
 # Permissions
 
-Documentation coming soon
+Permissions can be managed using `Permission` classes. A `Permission` class extends `BasePermission` and has a `has_permission` method. It can be hooked up to a field using the `permission_classes` keyword argument. A simple example looks like this:
```
Comment (jkimbo) [inline docs/guides/permissions.md:7]:
```suggestion
Permissions can be managed using `Permission` classes. A `Permission` class extends `BasePermission` and has a `has_permission` method. It can be added to a field using the `permission_classes` keyword argument. A simple example looks like this:
```

File: docs/guides/permissions.md [hunk 2]
```diff
@@ -4,4 +4,73 @@ title: Permissions
...
+
+    def has_permission(self, source: Any, info: Info, **kwargs) -> bool:
+        return False
+
+@strawberry.type
+class Query:
+    user: str = strawberry.field(permission_classes=[IsAuthenticated])
+```
+
+If the `has_permission` method fails then an error will be raised using the `message` class attribute. See [Dealing with Errors](/docs/guides/errors) for more information.
```
Comment (jkimbo) [inline docs/guides/permissions.md:25]:
```suggestion
If the `has_permission` method fails then an error will be raised using the `message` class attribute. See [Dealing with Errors](/docs/guides/errors) for more information on how errors are handled.
```

File: docs/guides/permissions.md [hunk 3]
```diff
@@ -4,4 +4,73 @@ title: Permissions
...
+    {
+      "message": "User is not authenticated"
+    }
+  ]
+}
+```
+
+In a simple use case you could obtain all the information you need to execute the `has_permission` check from the `source` and `info` objects.
+
+For more complex cases it might need more information from the Request. This can be provided by extending the `GraphQL` class to provide more context. The `Permission` class can then use that context from the `Info` object.
```
Comment (jkimbo) [inline docs/guides/permissions.md:40]: Most (maybe all?) of the views already pass the request as part of the context object. Wouldn't it be better to extract what you need in the permission class rather than modifying the context?

**Review by jkimbo — CHANGES_REQUESTED** [review]
Thanks for the contribution! I've added a couple of comments. Also I'm not sure the part explaining how to modify the context is necessary here.

File: docs/guides/permissions.md (hunk 3 above)
Comment (taybenlor) [reply docs/guides/permissions.md:40]: I didn't know that! I'll update the documentation.

In this case I'm basing the documentation off a solution I used to place a `user` object into the `context` (using the `Authorization` header of the request).

**Comment by taybenlor** [comment]
Thanks! I understand not wanting to introduce using the `context` object in this particular guide, but it was something I found useful while implementing. I think if not here, then there should be a specific guide on using it.

**Comment by taybenlor** [comment]
After trying to understand the properties available on the `context` object I think it would be very useful to document it.

**Comment by taybenlor** [comment]
I've finished up changes now and ready for re-review.

File: docs/guides/permissions.md [hunk 4]
```diff
@@ -4,4 +4,61 @@ title: Permissions
...
+
+class IsAuthenticated(BasePermission):
+    message = "User is not authenticated"
+
+    def has_permission(self, source: Any, info: Info, **kwargs) -> bool:
+        request: Union[Request, Websocket] = info.context["request"]
+        if "Authorization" in request.headers:
+          return authenticate_header(request)
+        if "auth" in request.query_params:
+          return authenticate_query_params(request)
```
Comment (jkimbo) [inline docs/guides/permissions.md:54]:
```suggestion
        if "Authorization" in request.headers:
            return authenticate_header(request)
        if "auth" in request.query_params:
            return authenticate_query_params(request)
```

File: docs/guides/permissions.md [hunk 5]
```diff
@@ -4,4 +4,61 @@ title: Permissions
...
+  "data": null,
+  "errors": [
+    {
+      "message": "User is not authenticated"
+    }
+  ]
+}
+```
+
+In most use cases you can obtain all the information you need to execute the `has_permission` check from the `source` and `info` objects. For example the `info` object contains a `context` dictionary, which includes the Request object. The Request object can be used to retrieve headers, query parameters, or cookies that can be used for auth like so:
```
Comment (jkimbo) [inline docs/guides/permissions.md:38]:
```suggestion

## Accessing user information

Accessing the current user information to implement your permission checks depends on the web framework you are using. Most frameworks will have a Request object where you can either access the current user directly or access headers/cookies/query parameters to authenticate the user. All the Strawberry integrations will provide this Request object in the `info.context` object that is accessible in every resolver and in the `has_permission` function. You can find more details about a specific framework integration under the "Integrations" heading in the navigation. 

In this example we are using `starlette` which uses the [ASGI](/docs/integrations/asgi) integration:
```

**Review by jkimbo — CHANGES_REQUESTED** [review]
I agree that documenting what is available on the context object is a good idea but I think that should go in the specific integration pages because they will differ slightly between them. Like the example you've provided for starlette as well. I've added some suggestions on how to reword it to make it clear that the specifics on how you implement the auth checks are going to differ between integrations. I don't want to imply that Strawberry always uses starlette for example.

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/),so make sure to pick the appropriate type. If in doubt feel free to ask :)

File: docs/guides/permissions.md (hunk 5 above)
Comment (taybenlor) [reply docs/guides/permissions.md:38]: Ahhh of course. I hadn't realised they would be different.

**Comment by taybenlor** [comment]
Fantastic, thanks for the suggestions. I'm happy with where that is now.

File: docs/guides/permissions.md [hunk 6]
```diff
@@ -4,4 +4,66 @@ title: Permissions
...
+
+    def has_permission(self, source: Any, info: Info, **kwargs) -> bool:
+        return False
+
+@strawberry.type
+class Query:
+    user: str = strawberry.field(permission_classes=[IsAuthenticated])
+```
+
+If the `has_permission` method fails then an error will be raised using the `message` class attribute. See [Dealing with Errors](/docs/guides/errors) for more information on how errors are handled.
```
Comment (BryceBeagle) [inline docs/guides/permissions.md:25]: Might want to be explicit on what `fails` means. Returns false? Throws an exception?

File: docs/guides/permissions.md [hunk 7]
```diff
@@ -4,4 +4,66 @@ title: Permissions
...
+      "message": "User is not authenticated"
+    }
+  ]
+}
+```
+
+
+## Accessing user information
+
+Accessing the current user information to implement your permission checks depends on the web framework you are using. Most frameworks will have a Request object where you can either access the current user directly or access headers/cookies/query parameters to authenticate the user. All the Strawberry integrations will provide this Request object in the `info.context` object that is accessible in every resolver and in the `has_permission` function. You can find more details about a specific framework integration under the "Integrations" heading in the navigation.
```
Comment (BryceBeagle) [inline docs/guides/permissions.md:41]:
```suggestion
Accessing the current user information to implement your permission checks depends on the web framework you are using. Most frameworks will have a `Request` object where you can either access the current user directly or access headers/cookies/query parameters to authenticate the user. All the Strawberry integrations provide this Request object in the `info.context` object that is accessible in every resolver and in the `has_permission` function. You can find more details about a specific framework integration under the "Integrations" heading in the navigation. 
```

Comment (BryceBeagle) [reply docs/guides/permissions.md:41]: > You can find more details about a specific framework integration under the "Integrations" heading in the navigation

Can we make this a link?

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
Can we add some line wrapping to each long line to match the style of the other docs in the repo?

**Comment by patrick91** [comment]
This is amazing! Thanks a lot for taking time to write this, let me know if @BryceBeagle and @jkimbo comments make sense 😊

File: docs/guides/permissions.md (hunk 7 above)
Comment (jkimbo) [reply docs/guides/permissions.md:41]: @BryceBeagle not at the moment. It's just a heading in the sidebar and not an actual page.

File: docs/guides/permissions.md (hunk 6 above)
Comment (taybenlor) [reply docs/guides/permissions.md]: Have cleared this up as `False` or falsy.

**Comment by taybenlor** [comment]
Thanks @patrick91!

@BryceBeagle have reworded and reformatted per review.

**Comment by 2malh** [comment]
I think the guide is missing the info that `has_permission` must
- a) contain some logic and 
- b) return a value other than False to properly execute without raising message.

To include Django as an example for `IsAuthenticated`
```py
def has_permission(self, source: Any, info: Info, **kwargs) -> bool:
        return info.context.request.user.is_authenticated
```

**Comment by taybenlor** [comment]
Thanks @2malh - have fixed this.

@BryceBeagle and @jkimbo your feedback has also been fixed.

**Review by jkimbo — APPROVED** [review]
Thanks @taybenlor

**Comment by taybenlor** [comment]
Cheers!
